zcash_client_sqlite/src/wallet/init.rs

```diff
@@ -1178,6 +1349,69 @@ mod tests {
         init_wallet_db(&mut db_data, Some(Secret::new(seed.to_vec()))).unwrap();
     }
 
+    #[test]
+    fn transaction_views() {
+        let data_file = NamedTempFile::new().unwrap();
+        let mut db_data = WalletDb::for_path(data_file.path(), tests::network()).unwrap();
+        init_wallet_db(&mut db_data, None).unwrap();
+
+        db_data.conn.execute_batch(
+            "INSERT INTO accounts (account, ufvk) VALUES (0, '');
+            INSERT INTO blocks (height, hash, time, sapling_tree) VALUES (0, 0, 0, '');
+            INSERT INTO transactions (block, id_tx, txid) VALUES (0, 0, '');
+            INSERT INTO sent_notes (tx, output_pool, output_index, from_account, address, value) VALUES (0, 2, 0, 0, '', 2);
+            INSERT INTO sent_notes (tx, output_pool, output_index, from_account, address, value) VALUES (0, 2, 1, 0, '', 3);
+            INSERT INTO received_notes (tx, output_index, account, diversifier, value, rcm, nf, is_change) VALUES (0, 0, 0, '', 5, '', 'a', false);
+            INSERT INTO received_notes (tx, output_index, account, diversifier, value, rcm, nf, is_change) VALUES (0, 1, 0, '', 7, '', 'b', true);",
+        ).unwrap();
+
+        let mut q = db_data
+            .conn
+            .prepare("SELECT received_total, received_note_count FROM v_tx_received")
+            .unwrap();
+        let mut rows = q.query(NO_PARAMS).unwrap();
+        let mut row_count = 0;
+        while let Some(row) = rows.next().unwrap() {
+            row_count += 1;
+            let total: i64 = row.get(0).unwrap();
+            let count: i64 = row.get(1).unwrap();
+            assert_eq!(total, 12);
+            assert_eq!(count, 2);
+        }
+        assert_eq!(row_count, 1);
+
+        let mut q = db_data
+            .conn
+            .prepare("SELECT sent_total, sent_note_count FROM v_tx_sent")
+            .unwrap();
+        let mut rows = q.query(NO_PARAMS).unwrap();
+        let mut row_count = 0;
+        while let Some(row) = rows.next().unwrap() {
+            row_count += 1;
+            let total: i64 = row.get(0).unwrap();
+            let count: i64 = row.get(1).unwrap();
+            assert_eq!(total, 5);
+            assert_eq!(count, 2);
+        }
+        assert_eq!(row_count, 1);
+
+        let mut q = db_data
+            .conn
+            .prepare("SELECT net_value, has_change FROM v_transactions")
+            .unwrap();
+        let mut rows = q.query(NO_PARAMS).unwrap();
+        let mut row_count = 0;
+        while let Some(row) = rows.next().unwrap() {
+            row_count += 1;
+            let net_value: i64 = row.get(0).unwrap();
+            let has_change: bool = row.get(1).unwrap();
+            assert_eq!(net_value, 7);
```

Comment:
Yep, this test shows the bug I raised above. This is ostensibly computing `5received + 7change - 2sent - 3sent`, but:
- if we treat all notes as separate, then we should have `0` instead of `7` because we want to ignore change (as it is wallet-internal, and the sum doesn't count the values of the wallet's own notes that were spent in the sent transactions).
- if the inconsistency is fixed, then we should also have `0` instead of `7` because we could not have any change note (as by definition change notes don't appear in `sent_notes`.

Reply:
This reminds me of a change that I'm making in the zcashd wallet, which I also want to make in the light wallet, which is that change be explicit, not implicit. The fact that change notes currently never show up as sent, only as received, means that we aren't really doing proper double-entry accounting in the wallet; we're eliding the change spent and then therefore have to skip counting the change received. Also, we're probably not accounting correctly for fees at all.